<a href="https://colab.research.google.com/github/qxou14/csc448-Spring2022/blob/main/Assignment_3.ipynb">
  <img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/>
</a>

In [8]:
# Imports section
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
import numpy as np
from sklearn.model_selection import cross_val_score
from sklearn import linear_model
from sklearn.preprocessing import PolynomialFeatures
from sklearn.pipeline import Pipeline

## Part 1. Loading the dataset

In [9]:
# Using pandas load the dataset (load remotely, not locally)
slimes = pd.read_csv("https://raw.githubusercontent.com/profmcnich/example_notebook/main/science_data_large.csv")


In [10]:
# Display a summary of the table information (number of datapoints, etc.)
slimes.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL        1000 non-null   int64  
 2   Size nm^3       1000 non-null   float64
dtypes: float64(1), int64(2)
memory usage: 23.6 KB


In [11]:
# Output the first 15 rows of the data
head = slimes.head(15)

head

Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04

## Part 2. Splitting the dataset

In [12]:
# Take the pandas dataset and split it into our features (X) and label (y)

x = slimes[['Temperature °C','Mols KCL']]
y = slimes['Size nm^3']


In this case x will be independent varaible and y will be dependent varaible.  Both tempeature and KCL will be independent, while size depends on them.

In [13]:
# this is x
x

Temperature °C  Mols KCL
0               469       647
1               403       694
2               302       975
3               779       916
4               901        18
..              ...       ...
995             894       847
996             327       982
997             791       213
998             769       553
999             919       452

[1000 rows x 2 columns]

In [14]:
# this is y
y

0      6.244743e+05
1      5.779610e+05
2      6.196847e+05
3      1.460449e+06
4      4.325726e+04
           ...     
995    1.545661e+06
996    6.737041e+05
997    3.477543e+05
998    8.684794e+05
999    8.476413e+05
Name: Size nm^3, Length: 1000, dtype: float64

In [15]:
# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)

x_train, x_test, y_train, y_test = train_test_split(x,y,test_size=0.1,train_size=0.9,random_state=42)

below will make sure the code works

In [16]:
#train x 
x_train

Temperature °C  Mols KCL
716              95       499
351             243       384
936              47       211
256             495       497
635              82       421
..              ...       ...
106             439        40
270             665       245
860             471       620
435             829       259
102             636       388

[900 rows x 2 columns]

In [17]:
#test x
x_test

Temperature °C  Mols KCL
521             100       541
737             635       668
740             799       665
660             966       871
411             785       595
..              ...       ...
436             421       725
764             431       860
88              983       433
63              580       817
826              24        89

[100 rows x 2 columns]

In [18]:
#train y  
y_train

716    103064.31430
351    193753.02860
936     21670.02857
256    505027.40000
635     75092.02857
           ...     
106     40433.71429
270    335545.00000
860    600674.85710
435    441286.60000
102    505469.25710
Name: Size nm^3, Length: 900, dtype: float64

In [19]:
y_test

521    1.177623e+05
737    8.687293e+05
740    1.084893e+06
660    1.716039e+06
411    9.536850e+05
           ...     
436    6.305199e+05
764    7.676234e+05
88     8.684308e+05
63     9.737511e+05
826    4.786314e+03
Name: Size nm^3, Length: 100, dtype: float64

## Part 3. Perform a Linear Regression

In [20]:
# Use sklearn to train a model on the training set
reg = LinearRegression().fit(x_train, y_train)

In [21]:
reg.score(x,y)

0.8606369652963943

In [22]:
reg.intercept_

-409391.47958340775

In [23]:
#i will use this data to test the expected  answer for this is   431, 860,   ->  767623.4286
predict_data= x_test[x_test["Temperature °C"] == 431]


In [24]:
predict_data 

Temperature °C  Mols KCL
764             431       860

In [25]:

# Create a sample datapoint and predict the output of that sample with the trained model
#This is taken from one of the test input. I copied the expected answer at the end.
reg.predict(predict_data)

array([852035.38175768])

The expected answer is 767623.4286 but we got 852035.38175768. Quite off. :( . However, we have so many points with different data. It is doing its best with best fit line. I strongly believe our results will be improved if there are more trainning sets.  But maybe it is very accruate for other inputs.

In [26]:
# Report on the score for that model, in your own words (markdown, not code) explain what the score means

scores = reg.score(x_test, y_test, sample_weight=None)
scores

0.8552472077276095

Wow ! That's why I was still confident about the model even though the last testing is quite off.  As we can see, we get approximitely a score of ~0.85.   The best posssible score for this is 1.0, this means our model is 85% accurate, which is quite accurate considering we only have 900 trainning sets.

In [27]:
# Extract the coefficents and intercept from the model and write an equation for your h(x) using LaTeX
coef = reg.coef_
intercept = reg.intercept_

print(coef)
print(intercept)

[ 866.14641337 1032.69506649]
-409391.47958340775


$ h(x) = 866x_1 + 1033x_2 - 409391 $

For the above equations I remove all the digits to make this much cleaner to read. (The instruction also states I can do it this way)

## Part 4. Use Cross Validation

In [28]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data

lasso = linear_model.Lasso()
#default cv is 5
cross_score = cross_val_score(lasso, x, y)
print(cross_score)




[0.83918826 0.87051239 0.85871066 0.87202623 0.84364641]


### Report on their finding and their significance

According to wikipedia, cross validation is a way that uses different portion of the data as trainning sets and testing sets for a model. In this case, we can use this to test the correctness of our model.

Based on the results that we got from cross validation, we got 0.83918826 0.87051239 0.85871066 0.87202623 0.84364641. All of these are very similar to the score we got before which is 0.8552472077276095. As a result, model that we got is pretty good. The result we got is pretty close to 1.0 and we can reproduce simiar result by using cross validation mutiple times.

## Part 5. Using Polynomial Regression

In [29]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
poly = PolynomialFeatures(degree=2)
poly_transform = poly.fit_transform(x)

#now x is transform into polunomials that we can apply in linear models
#we will repeat the process 
poly_x_train, poly_x_test, poly_y_train, poly_y_test = train_test_split(poly_transform,y,test_size=0.1,train_size=0.9,random_state=42)

In [30]:
reg_poly = LinearRegression().fit(poly_x_train, poly_y_train)


In [31]:
poly_coef = reg_poly.coef_
poly_intercept = reg_poly.intercept_

print(poly_coef)
print(poly_intercept)

[ 0.00000000e+00  1.20000000e+01 -1.27196718e-07  1.26494371e-11
  2.00000000e+00  2.85714287e-02]
2.047873567789793e-05


based on the documentation, we are converting $[ x_1 , x_2]$ to $[1,x_1,x_2,x^2,x_1x_2,x_2^2]$


### Report on the metrics and output the resultant equation as you did in Part 3.

h(x) = 2.04 $*$ ${10^{-5}}$ + 0 + 1.2 $*$ $10^1x_1$ - 1.27 $*$ ${10^{-7}}$ $x_2$ + 1.26 $*$ ${10^{-11}}$ $x^2$ + 2 $*$ $x_1x_2$ + 2.85 $*$ ${10^{-2}}$ $x_2^2$




If we simply this and ignore some very small number, we will get

$ h(x) $ = 0.0000204 + 12$x_1$   + 2$x_1x_2$ + 0.0285 ${x_2^2}$ 